router/data_store.go

```diff
@@ -39,196 +33,206 @@ type SyncHandler interface {
+	return &pgDataStore{
+		pgx:   pgx,
+		rt:    routeType,
+		donec: make(chan struct{}),
+	}
 }
 
-var ErrExists = errors.New("router: route already exists")
-var ErrNotFound = errors.New("router: route not found")
+var sqlAddRoute = `INSERT INTO routes (parent_ref, type, config) VALUES ($1, $2, $3::json) RETURNING route_id, created_at, updated_at`
 
-func (s *etcdDataStore) Add(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+func (d *pgDataStore) Add(r *router.Route) error {
+	if r.CreatedAt == nil {
+		r.CreatedAt = new(time.Time)
 	}
-	_, err = s.etcd.Create(s.path(r.ID), string(data), 0)
-	if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 105 {
-		err = ErrExists
+	if r.UpdatedAt == nil {
+		r.UpdatedAt = new(time.Time)
 	}
-	return err
+	return d.pgx.QueryRow(sqlAddRoute, r.ParentRef, r.Type, r.Config).Scan(&r.ID, r.CreatedAt, r.UpdatedAt)
 }
 
-func (s *etcdDataStore) Set(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+var sqlSetRoute = `UPDATE routes SET parent_ref = $1, type = $2, config = $3::json WHERE route_id = $4 RETURNING updated_at`
+
+func (d *pgDataStore) Set(r *router.Route) error {
+	if r.UpdatedAt == nil {
+		r.UpdatedAt = new(time.Time)
+	}
+	err := d.pgx.QueryRow(sqlSetRoute, r.ParentRef, r.Type, r.Config, r.ID).Scan(r.UpdatedAt)
+	if err == pgx.ErrNoRows {
+		return ErrNotFound
 	}
-	_, err = s.etcd.Set(s.path(r.ID), string(data), 0)
 	return err
 }
 
-func (s *etcdDataStore) Remove(id string) error {
-	_, err := s.etcd.Delete(s.path(id), true)
-	if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 100 {
-		return ErrNotFound
-	}
+var sqlRemoveRoute = `UPDATE routes SET deleted_at = now() WHERE route_id = $1`
+
+func (d *pgDataStore) Remove(id string) error {
+	_, err := d.pgx.Exec(sqlRemoveRoute, id)
 	return err
 }
 
-func (s *etcdDataStore) Get(id string) (*router.Route, error) {
-	res, err := s.etcd.Get(s.path(id), false, false)
+var sqlGetRoute = `SELECT parent_ref, type, config, created_at, updated_at FROM routes WHERE route_id = $1 AND type = $2 AND deleted_at IS NULL`
+
+func (d *pgDataStore) Get(id string) (*router.Route, error) {
+	r := &router.Route{ID: id, Config: new(router.Config), CreatedAt: new(time.Time), UpdatedAt: new(time.Time)}
+	row := d.pgx.QueryRow(sqlGetRoute, id, d.rt)
+
+	err := row.Scan(&r.ParentRef, &r.Type, r.Config, r.CreatedAt, r.UpdatedAt)
+	if err == pgx.ErrNoRows {
+		return nil, ErrNotFound
+	}
 	if err != nil {
-		if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 100 {
-			err = ErrNotFound
-		}
 		return nil, err
 	}
-	r := &router.Route{}
-	err = json.Unmarshal([]byte(res.Node.Value), r)
-	return r, err
+
+	return r, nil
 }
 
-func (s *etcdDataStore) List() ([]*router.Route, error) {
-	res, err := s.etcd.Get(s.prefix, false, true)
+var sqlListRoutes = `SELECT route_id, parent_ref, type, config, created_at, updated_at FROM routes WHERE type = $1 AND deleted_at IS NULL`
```

Comment:
This query probably needs an index, but we can skip it if the table split is coming next.